File: backend/shared/models/reconciliation.py
"""Reconciliation models."""
from datetime import datetime
from enum import Enum
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


class ReconciliationStatus(str, Enum):
    """Reconciliation processing status."""

    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


class PartnerAllocation(BaseModel):
    """Allocation result for a single partner."""

    partner_id: str
    partner_name: str
    gross_volume: float
    bsw_percent: float
    water_cut_factor: float
    net_volume_observed: float
    temperature_correction_factor: float
    api_correction_factor: float
    net_volume_standard: float
    ownership_percent: float
    allocated_volume: float
    intermediate_calculations: Dict[str, Any] = Field(default_factory=dict)


class ReconciliationBase(BaseModel):
    """Base reconciliation fields."""

    period_start: datetime
    period_end: datetime
    terminal_volume: float = Field(gt=0, description="Final terminal volume in barrels")


class ReconciliationCreate(ReconciliationBase):
    """Reconciliation creation model."""

    tenant_id: str
    triggered_by: str


class ReconciliationResult(BaseModel):
    """Detailed reconciliation results."""

    total_gross_volume: float
    total_net_volume_standard: float
    total_allocated_volume: float
    shrinkage_volume: float
    shrinkage_percent: float
    partner_allocations: List[PartnerAllocation]
    allocation_model_used: str


class Reconciliation(ReconciliationBase):
    """Reconciliation model with metadata."""

    id: str
    tenant_id: str
    triggered_by: str
    status: ReconciliationStatus = ReconciliationStatus.PENDING
    result: Optional[ReconciliationResult] = None
    error_message: Optional[str] = None
    created_at: datetime
    completed_at: Optional[datetime] = None

    class Config:
        from_attributes = True

File: frontend/src/types/reconciliation.ts
export enum ReconciliationStatus {
  PENDING = "pending",
  PROCESSING = "processing",
  COMPLETED = "completed",
  FAILED = "failed",
}

export interface PartnerAllocation {
  partner_id: string
  partner_name: string
  gross_volume: number
  bsw_percent: number
  water_cut_factor: number
  net_volume_observed: number
  temperature_correction_factor: number
  api_correction_factor: number
  net_volume_standard: number
  ownership_percent: number
  allocated_volume: number
  intermediate_calculations?: Record<string, any>
}

export interface ReconciliationResult {
  total_gross_volume: number
  total_net_volume_standard: number
  total_allocated_volume: number
  shrinkage_volume: number
  shrinkage_percent: number
  partner_allocations: PartnerAllocation[]
  allocation_model_used: string
}

export interface Reconciliation {
  id: string
  tenant_id: string
  triggered_by: string
  period_start: string
  period_end: string
  terminal_volume: number
  status: ReconciliationStatus
  result?: ReconciliationResult
  error_message?: string
  created_at: string
  completed_at?: string
}

export interface TerminalReceipt {
  id: string
  tenant_id: string
  receipt_date: string
  terminal_volume: number
  terminal_name?: string
  operator_name?: string
  notes?: string
  created_by: string
  created_at: string
  updated_at?: string
}

export interface PaginationMeta {
  current_page: number
  page_size: number
  total_count: number
  total_pages: number
  has_next: boolean
  has_previous: boolean
}

export interface PaginatedTerminalReceipts {
  data: TerminalReceipt[]
  pagination: PaginationMeta
}
